platform/broadcom/sonic-platform-modules-ufispace/s9300-32d/sonic_platform/fan.py

```diff
@@ -0,0 +1,181 @@
+        if self.is_psu_fan:
+            attr = "psu_fan{}_speed_rpm".format(self.fan_index)
+            device = "PSU{}".format(self.fans_psu_index)
+            max_speed = int(self.plugin_data['PSU']['PSU_FAN_MAX_SPEED']) 
+        else:
+            if self.fan_index == 1:
+                pos = "f"
+                max_speed = int(self.plugin_data['FAN']['FAN_F_MAX_SPEED']) 
+            else:
+                pos = "r"
+                max_speed = int(self.plugin_data['FAN']['FAN_R_MAX_SPEED']) 
+            attr = "fan{}_{}_speed_rpm".format(self.fantray_index, pos)
+            device = "FAN-CTRL"
+
+        output = self.pddf_obj.get_attr_name_output(device, attr)
+        if not output:
+            return speed_percentage
+
+        output['status'] = output['status'].rstrip()
+        if output['status'].isalpha():
+            return speed_percentage
+        else:
+            speed = int(float(output['status']))
+
+        speed_percentage = round((speed*100)/max_speed)
+
+        if speed_percentage > 100:
+            speed_percentage = 100
+        
+        return speed_percentage
+
+    def get_speed_rpm(self):
+        """
+        Retrieves the speed of fan in RPM
+
+        Returns:
+            An integer, Speed of fan in RPM
+        """
+        rpm_speed = 0
+        if self.is_psu_fan:
+            attr = "psu_fan{}_speed_rpm".format(self.fan_index)
+            device = "PSU{}".format(self.fans_psu_index)
+        else:
+            if self.fan_index == 1:
+                pos = "f"
+            else:
+                pos = "r"
+            attr = "fan{}_{}_speed_rpm".format(self.fantray_index, pos)
+            device = "FAN-CTRL"
+
+        output = self.pddf_obj.get_attr_name_output(device, attr)
+
+        if output is None:
+            return rpm_speed
+
+        output['status'] = output['status'].rstrip()
+        if output['status'].isalpha():
+            return rpm_speed
+        else:
+            rpm_speed = int(float(output['status']))
+
+        return rpm_speed
+
+    def get_direction(self):
+        """
+        Retrieves the direction of fan
+        Returns:
+            A string, either FAN_DIRECTION_INTAKE or FAN_DIRECTION_EXHAUST
+            depending on fan direction
+        """
+        direction = self.FAN_DIRECTION_INTAKE
+        if self.is_psu_fan:
+            attr = "psu_fan{}_dir".format(self.fan_index)
+            device = "PSU{}".format(self.fans_psu_index)
+        else:
+            attr = "fan{}_dir".format(self.fantray_index)
+            device = "FAN-CTRL"
+
+        output = self.pddf_obj.get_attr_name_output(device, attr)
+        if not output:
+            return direction
+
+        mode = output['mode']
+        val = output['status'].strip()
+        vmap = self.plugin_data['FAN']['direction'][mode]['valmap']
+   
+        if val in vmap:
+            direction = vmap[val]
+            
+        return direction
+
+    def get_presence(self):
```

Comment:
This implementation is pretty similar to the one present in pddf_fan.py. Can you double check why you had to overwrite that def with this one?

Reply:
Currently, the base always set the psu fan presence to true which is FRU in our platform. So we overwrite it to dectect PSU presence as for psu fan presence.